content/blog/2023-12-11-BrilCpp.md

```diff
@@ -0,0 +1,47 @@
++++
+title = "C++ Infrastructure for Bril"
+[[extra.authors]]
+name = "Albert Xiao"
+[[extra.authors]]
+name = "Ryan Mao"
+[extra]
+bio="""
+Albert Xiao is an M.Eng student studying CS. He is particularly interested in systems and compilers development. 
+Ryan Mao is an M.Eng student studying CS. He is also interested in compilers.
+"""
++++
+
+## C++ Infrastructure for Bril
+Albert Xiao & Ryan Mao
+
+### Goal
+Our primary aim was to enhance Bril's capabilities by incorporating a C++ interface, equipped with a parser, JSON printer, optimized types for instructions, and streamlined program flow mutations. The focus was squarely on performance, user-friendliness, and the potential to expand the Bril ecosystem with lightning-fast C++ optimizations.
+
+### Approach
+We already built a basic C++ interface for our use in the assignments and exercises throughout the course. For our final project, we used this interface as a starting point, and aimed to build off of it to create a successful framework for Bril. First and foremost, we identified critical areas of our interface for improvement: shortcomings in memory safety, performance, and user-friendliness. Our strategy was twofold: retain certain functional aspects while fundamentally revamping the framework to align with our objectives.
+
+Our initial phase involved a meticulous redesign of Bril's program types within our framework. We restructured our infrastructure around control flow graphs (CFG), because most optimizations operate at this level. This involved functionality to divide functions into basic blocks, and equipping program types with hooks to store analysis information for later optimization. We integrated data encapsulation into our types, in order to ensure memory safety and resistance to implementation changes.
```

Comment:
What does this "data encapsulation" look like? What were the old (poorly encapsulated) types, and what are their improved equivalents?